### PR TITLE
Enhance `yii\rbac\ManagerInterface::getUserIdsByRole()`

### PR DESCRIPTION
**Pull request by Kolyunya**
| Q | A |
| --- | --- |
| Is bugfix? | no |
| New feature? | yes |
| Breaks BC? | no |
| Tests pass? | yes |
| Fixed issues | #11895 |

It is now possible to get users by role recursively.

### CODE REVIEW

File: framework/rbac/DbManager.php [hunk 1]
```diff
@@ -753,6 +754,24 @@ public function hasChild($parent, $child)
     /**
      * @inheritdoc
      */
+    public function getParents($name)
```
Comment (pstddc) [inline framework/rbac/DbManager.php]: may be add second argument `getParents($name, $type = Item::TYPE_ROLE)` and add condition for select?

**Comment by Kolyunya** [comment]
@samdark I would be glad to make any improvements to this patch if you provide me with information about it's issues.

File: framework/rbac/DbManager.php (hunk 1 above)
Comment (samdark) [inline framework/rbac/DbManager.php]: Yes, that could be useful.

File: framework/rbac/ManagerInterface.php [hunk 2]
```diff
@@ -161,6 +161,13 @@ public function removeChildren($parent);
     public function hasChild($parent, $child);
 
     /**
+     * Returns the parent permissions and/or roles.
+     * @param string $name the child name
+     * @return Item[] the parent permissions and/or roles
+     */
```
Comment (samdark) [inline framework/rbac/ManagerInterface.php]: Need `@since 2.0.10`

File: framework/rbac/DbManager.php [hunk 3]
```diff
@@ -966,18 +985,36 @@ public function loadFromCache()
     /**
      * Returns all role assignment information for the specified role.
      * @param string $roleName
+     * @param string $recursive
```
Comment (samdark) [inline framework/rbac/DbManager.php]: Need to elaborate what is recursive and how it works.

File: framework/rbac/ManagerInterface.php [hunk 4]
```diff
@@ -212,10 +219,11 @@ public function getAssignments($userId);
     /**
      * Returns all user IDs assigned to the role specified.
      * @param string $roleName
+     * @param string $recursive
```
Comment (samdark) [inline framework/rbac/ManagerInterface.php]: Need description.

File: framework/rbac/DbManager.php (hunk 1 above)
Comment (Kolyunya) [inline framework/rbac/DbManager.php]: This would be inconsistent with `getChildren()` and require its refactoring too.

Comment (samdark) [inline framework/rbac/DbManager.php]: OK.

**Comment by Kolyunya** [comment]
@samdark this feature doesn't break the BC. Why can't we have it in `2.0.x`?

**Comment by dynasource** [comment]
@Kolyunya, thats because you are changing an interface signature at https://github.com/yiisoft/yii2/pull/11927/files#diff-72766818d5fa80f81ef61a9d0dde4b63L218

**Comment by Kolyunya** [comment]
@dynasource I do change the signature but it's compatible with the previous version and can not break the backward compatibility.

**Comment by samdark** [comment]
The interface change is actually backwards compatible.

**Comment by Kolyunya** [comment]
@samdark `ManagerInterface` changed. Custom implementations will break. How is it backward compatible?

**Comment by samdark** [comment]
https://3v4l.org/VSVdn — default value is a key.

**Comment by Kolyunya** [comment]
@samdark **custom** managers will break after **interface** change, won't they?

**Comment by samdark** [comment]
They won't if `$recursive` value has a default.

**Comment by Kolyunya** [comment]
@samdark could you provide an example? It seems, [it's not going to work out](https://3v4l.org/A48Rh).

**Comment by samdark** [comment]
Reduction of parameters doesn't work like that. Adding more parameters with default values does.

**Comment by samdark** [comment]
So yeah. You're right.

**Pull request closed without merging**